appsembler_reporting role

bryanlandia opened the pull request:
This adds an appsembler_reporting role to create the shell scripts, cron jobs, and log dirs to run and log the reports.

Installs the edxapp server; or, in the case of enterprise customers, to edxapp server 1.

bryanlandia committed:
appsembler_reporting role initial commit:
Make reporting shell scripts, cron jobs to run, and log dir for output

Bug fixes on appsembler_reporting role

bryanlandia commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Trinity: reports didn't get generated in Appsembler reporting](https://trello.com/c/McOunA9Y/3017-trinity-reports-didnt-get-generated-in-appsembler-reporting)

OmarIthawi commented on playbooks/roles/appsembler_reporting/tasks/main.yml:
```diff
@@ -0,0 +1,27 @@
+---
+
+- name: make logs directory for appsembler_reporting
```
@bryanlandia I don't think there's a need for `.sh` files here. We can directly create cronjobs as have been done in:

 - https://github.com/appsembler/edx-configs/blob/master/easycare/prod/files/customer-post.yml

bryanlandia replied:
@OmarIthawi Arrgh, I hate repeating work that's already been done :(  Yes, agreed about the shell scripts—I was just decided to replicate what @johnbaldwin had set up on ASU and keep them.  Your playbook is cleaner, but as this is something which will be replicated for multiple customers I like having a role.  What do you think?

OmarIthawi replied:
@bryanlandia work duplication is inevitable. The other, worse alternative is spamming everyone about everything you do. So please don't worry about it.

> as this is something which will be replicated for multiple customers I like having a role. What do you think?

You're right, having a playbook is much better than copying the file to every customer. So, please copy as much you'd like from `easycare/prod/files/customer-post.yml` and make it a proper role that we can reuse for others.

`customer-post.yml` should be used for customer-specific changes and hacks.

bryanlandia replied:
@OmarIthawi Great.  Yes, let's use `customer-post.yml` like for Trinity's case where there is (I think) an additional report to run, but everything else in the role.  I'll update the role here.

OmarIthawi reviewed (commented):
Thanks @bryanlandia! Could you please take a look at the attached playbook?

bryanlandia commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Trinity: discuss what to do about reports](https://trello.com/c/okKBOFmg/2941-trinity-discuss-what-to-do-about-reports)

bryanlandia commented:
As far as I can tell the failing CI job doesn't have anything to do with the new code and is probably a fluke.  We'll see what happens next time.

bryanlandia committed:
appsembler reporting: fix course metrics name, simplify config
removing j2 template intermediary for script creation, put scripts
in own dir.  (thanks @omarithawi)

bryanlandia commented:
@OmarIthawi Thanks for your help here.  I refactored the role using some of your approach and some of mine.  Please take another look.

OmarIthawi commented on playbooks/appsemblerPlaybooks/ficus_basic.yml:
```diff
@@ -81,3 +81,9 @@
       tags: ['ecommerce_theme_role']
+    - role: appsembler_reporting
+      when: 
+        - "EDXAPP_APPSEMBLER_FEATURES is defined and EDXAPP_APPSEMBLER_FEATURES.get('ENABLE_APPSEMBLER_REPORTING', False)"
```
-> `- EDXAPP_APPSEMBLER_FEATURES.EDXAPP_APPSEMBLER_FEATURES|default(false)`

OmarIthawi commented on playbooks/appsemblerPlaybooks/ficus_enterprise.yml:
```diff
@@ -137,3 +137,12 @@
+  when: 
+    - "EDXAPP_APPSEMBLER_FEATURES is defined and EDXAPP_APPSEMBLER_FEATURES.get('ENABLE_APPSEMBLER_REPORTING', False)"
+    - "COMMON_ENVIRONMENT == 'prod'"
+  tags: ['appsembler_reporting_role']
```
This doesn't look right. Please double check, as there's no `roles` nor `tasks`.

OmarIthawi commented on playbooks/roles/appsembler_reporting/defaults/main.yml:
```diff
@@ -0,0 +1,14 @@
+    cron_m: 0
+  learners:
+    cron_h: 2
+    cron_m: 0
```
Nip Tuck: A newline here, please.

OmarIthawi commented on playbooks/roles/appsembler_reporting/defaults/main.yml:
```diff
@@ -0,0 +1,14 @@
+    cron_m: 0
+  courses:
+    cron_h: 1
+    cron_m: 0
```
Please favor the following format as it's more Ansible-ic:

```
appsembler_reporting_jobs:
  - name: grades
    cron_h: 0
    cron_m: 0
  - name: courses
    cron_h: 1
    cron_m: 0
```

So I don't have to wonder what `item.value` means.

OmarIthawi commented on playbooks/roles/appsembler_reporting/tasks/main.yml:
```diff
@@ -0,0 +1,39 @@
+    dest: "{{ appsembler_reporting_scripts_dir }}/{{ item.key }}_report.sh"
+    owner: edxapp
+    group: edxapp
+    mode: 774
```
`774` doesn't make sense here. You could do it `770` and it would better. As you're giving `4` (read) to `others` and but not `1` (execute) so it's useless.

OmarIthawi commented on playbooks/roles/appsembler_reporting/tasks/main.yml:
```diff
@@ -0,0 +1,39 @@
+    path: "{{ appsembler_reporting_scripts_dir }}"
+    state: directory
+    owner: edxapp
+    group: edxapp    
```
Please specify a mode here. I'm thinking of `0775`, but not completely sure about it.

OmarIthawi commented on playbooks/roles/appsembler_reporting/tasks/main.yml:
```diff
@@ -0,0 +1,39 @@
+    hour: "{{ item.value.cron_h }}" # Hour of the day
+    minute: "{{ item.value.cron_m }}" # Minute of the hour
+    job: "bash -e {{ appsembler_reporting_scripts_dir }}/{{ item.key }}_report.sh"
+  with_dict: "{{ appsembler_reporting_jobs }}"
```
`with_dict` -> `with_items`, after addressing the earlier comment about `appsembler_reporting_jobs`.

OmarIthawi reviewed (commented):
Thanks @bryanlandia! Much better. I have 7 small nip tucks.

bryanlandia committed:
appsembler_reporting role fix conditionals, modes, ficus enterprise include

bryanlandia commented:
@OmarIthawi I pushed changes based on your recommendations.  Thanks!

OmarIthawi committed:
Fixup ficus enterprise reporting check

OmarIthawi reviewed (approved):
Thanks @bryanlandia. LGTM.
